incubator/hnc/internal/reconcilers/hierarchy_config.go

```diff
@@ -267,6 +274,33 @@ func (r *HierarchyConfigReconciler) syncExternalNamespace(log logr.Logger, nsIns
 	ns.ExternalTreeLabels = etls
 }
 
+// upgradeSubnamespaceAnnotation replaces the deprecated "subnamespaceOf"
+// annotation with the new "subnamespace-of" annotation. If both exist, we will
+// only delete the deprecated annotation.
+func (r *HierarchyConfigReconciler) upgradeSubnamespaceAnnotation(inst *corev1.Namespace) {
+	sOf, exist := inst.GetAnnotations()[api.DeprecatedSubnamespaceOf]
+	if !exist {
+		return
+	}
+
+	// Remove old annotation
+	a := inst.GetAnnotations()
+	delete(a, api.DeprecatedSubnamespaceOf)
+	inst.SetAnnotations(a)
+
+	// Add new annotation key with the old value if it doesn't exist or this is
+	// the first time this namespace gets reconciled, which is a corner case if
```

Comment:
I don't understand this "first time the namespace gets reconciled." Why should it matter if it's the first time or not? Generally speaking, the reconciler should be as stateless as possible and we shouldn't care what's come before (the only exception I know of is forcing _other_ objects to be re-reconciled once we have more information).

Also, what happens if this is the second time we've reconciled this namespace, but there was an error the first time we reconciled it and the namespace was never written back to the apiserver?

Reply:
RE second time reconciling: the complicated logic is mostly for lowering the risk at API conversion (when HNC restarts).

SGTM, updated the logic to always respect the new annotation.